src/intern.rs

```diff
@@ -506,6 +518,112 @@ mod intern_tests {
+        assert_eq!(interned_size, string_size + set_size + pointers_in_set_size);
+    }
+
+    #[cfg(feature = "deepsize")]
+    #[derive(Clone)]
+    struct ArcInside {
+        hash: usize,
+        pointer: Arc<String>,
+    }
+
+    #[cfg(feature = "deepsize")]
+    impl Hash for ArcInside {
+        /// For testing purposes, we only hash the hash field.
+        /// In order to make [`ArcInside`] instances containing the same string have different hash values.
+        fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+            self.hash.hash(state);
+        }
+    }
+
+    #[cfg(feature = "deepsize")]
+    impl PartialEq for ArcInside {
+        /// For testing purposes, we only compare the hash field.
+        fn eq(&self, other: &Self) -> bool {
+            self.hash == other.hash
+        }
+    }
+
+    #[cfg(feature = "deepsize")]
+    impl Eq for ArcInside {}
+
+    #[cfg(feature = "deepsize")]
+    impl DeepSizeOf for ArcInside {
+        fn deep_size_of_children(&self, context: &mut Context) -> usize {
+            self.pointer.deep_size_of_children(context)
+        }
+    }
+
+    #[cfg(feature = "deepsize")]
+    #[test]
+    fn test_deep_size_with_context() {
+        let string = String::from("abcdefghijklmnopqrstuvwxyz");
+        // size of string inside arc, 50 bytes.
+        // Three arcs pointed to the same string will not be counted multiple times.
+        let string_size = string.deep_size_of();
+
+        let a1 = ArcInside {
+            hash: 1,
+            pointer: Arc::new(string),
+        };
+        let a2 = ArcInside {
+            hash: 2,
+            pointer: a1.pointer.clone(),
+        };
+        let a3 = ArcInside {
+            hash: 3,
+            pointer: a1.pointer.clone(),
+        };
+        // size of ArcInside, 16 bytes each
+        let object_size = size_of::<ArcInside>() * 3;
+
+        let _ = Intern::new(a1);
+        let _ = Intern::new(a2);
+        let _ = Intern::new(a3);
+
+        // size of set
+        let set_size = INTERN_CONTAINERS.with(|m: &mut HashSet<&'static ArcInside>| size_of_val(m));
+        // size of pointers in the set
+        let pointers_in_set_size = INTERN_CONTAINERS.with(|m: &mut HashSet<&'static ArcInside>| {
+            size_of::<&'static ArcInside>() * m.capacity()
+        });
+
+        // 3 ArcInside has different hash values
+        INTERN_CONTAINERS.with(|m: &mut HashSet<&'static ArcInside>| assert_eq!(m.len(), 3));
+
+        let interned_size = deep_size_of_interned::<ArcInside>();
+        assert_eq!(
+            interned_size,
+            string_size + object_size + set_size + pointers_in_set_size
+        );
+
+        println!("string_size: {}", string_size);
+        println!("object_size: {}", object_size);
+        println!("set_size: {}", set_size);
+        println!("pointers_in_set_size: {}", pointers_in_set_size);
+        println!("interned_size: {}", interned_size);
```

Comment:
I'd prefer to put the printlns before the assertion, so if it fails it's easier to see what went wrong.
```suggestion
        println!("string_size: {}", string_size);
        println!("object_size: {}", object_size);
        println!("set_size: {}", set_size);
        println!("pointers_in_set_size: {}", pointers_in_set_size);
        println!("interned_size: {}", interned_size);

        assert_eq!(
            interned_size,
            string_size + object_size + set_size + pointers_in_set_size
        );
```